### PR TITLE
Fix tie breaking in ndcg metric

### LINKED ISSUE #2022: Self-Implemented Metrics Behaved Weirdly in DDP Mode
## 🐛 Bug

Self-implemented nDCG@5 is larger than nDCG@1 in DDP mode. What I was trying to do here is to record the sum of sample-wise scores **_s_** and the number of samples **_n_**. At the end of batch/epoch, I can obtain the mean of nDCG@k by dividing **_s_** by **_n_**.

My implementation looks very similar to the example on [IMPLEMENTING A METRIC](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html). However, It failed in distributed settings.

`Epoch 8:  60%|██████    | 1541/2555 [04:58<03:16,  5.17it/s, loss=0.00354, v_num=213259, train_ndcg@1=0.656, train_ndcg@3=0.675, train_ndcg@5=0.698,...`

### To Reproduce

<details>
  <summary>Self-implemented Metric</summary>

```python
from torchmetrics.functional.retrieval import retrieval_normalized_dcg
from torchmetrics import Metric
import torch
from torch import Tensor

class NDCG(Metric):
    full_state_update = False

    def __init__(self, top_k: int):
        super().__init__()
        self.top_k = top_k
        self.add_state("ndcg", default=torch.tensor(0, dtype=torch.float), dist_reduce_fx="sum")
        self.add_state("n", default=torch.tensor(0), dist_reduce_fx="sum")

    def update(self, preds: Tensor, target: Tensor):
        # preds and target are 2-D tensors.
        assert preds.shape == target.shape

        self.ndcg += torch.stack([retrieval_normalized_dcg(p, t, k=self.top_k) for p, t in zip(preds, target)]).sum()
        self.n += len(preds)

    def compute(self):
        return self.ndcg / self.n
```

</details>

### Expected behavior

The documentation of self-implemented metric looked too simple and probably missed some details of distributed metrics.

I noticed that you used list and in `compute()` you call mean() to get the average score of a given batch. Can you please explain the difference between these two implementations?

### Environment
- TorchMetrics version (and how you installed TM, e.g. conda, pip, build from source):
0.10.3 (conda-forge)
- Python version:
3.10.12
PyTorch version (e.g., 1.0):
1.13.1
- Any other relevant information such as OS (e.g., Linux):
pytorch-lightning: 1.9.4
Linux: 5.15.0-78-generic

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2022
In the issue it was identified that the overall problem is that our implementation of ndcg currently cannot deal with tie breaks of scores. PR fixes this.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2031.org.readthedocs.build/en/2031/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by SkafteNicki** [comment]
@lucadiliello could you please review as you are the expert on retrieval?

**Comment by lucadiliello** [comment]
I think this issue may also affect other retrieval metrics. The problem with other retrieval metrics is that definitions are not really standardized between Wikipedia and slides of university courses. I mean that the general definition is the same, but the behavior in corner cases like this one is not well defined.

I'm currently on vacation, I will try to review the PR once I'm in front of my pc.

**Comment by Borda** [comment]
> I'm currently on vacation, I will try to review the PR once I'm in front of my pc.

cool, let us know when you ll check and and ideally if all fine approve this PR :)

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@lucadiliello pls mind checking it ^^

**Review by lucadiliello — APPROVED** [review]

**Comment by lucadiliello** [comment]
LGTM
